components/action-mgt/org.wso2.carbon.identity.action.execution/src/main/java/org/wso2/carbon/identity/action/execution/util/APIClient.java

```diff
@@ -202,13 +217,24 @@ private void handleServerError(ActionInvocationResponse.Builder builder, HttpEnt
             builder.errorLog("Unexpected error response received for the status code: " + statusCode + ". "
                     + e.getMessage());
             builder.retry(true);
+            ActionInvocationErrorResponse invalidResponse = handleInvalidResponse();
+            builder.response(invalidResponse);
         }
     }
 
-    private ActionInvocationSuccessResponse handleSuccessResponse(HttpEntity responseEntity)
+    private ActionInvocationErrorResponse handleInvalidResponse() {
+
+        return new ActionInvocationErrorResponse.Builder()
+                .actionStatus(ActionInvocationResponse.Status.ERROR)
+                .error(ActionExecutionApiClientConstants.serverError)
+                .errorDescription(ActionExecutionApiClientConstants.serverErrorMessage)
+                .build();
+    }
+
+    private ActionInvocationResponse.APIResponse handleSuccessOrFailureResponse(HttpEntity responseEntity)
             throws ActionInvocationException {
 
-        return deserialize(responseEntity, ActionInvocationSuccessResponse.class);
+        return deserialize(responseEntity);
```

Comment:
Why we implementing new private method for this. Shall we check the `isSuccess` value here and use the existing deserialize method.